src/Common/ConcurrencyControl.h

```diff
@@ -0,0 +1,207 @@
+            parent.schedule(lock);
+
+            std::unique_lock lock2{mutex};
+            released++;
+            assert(released <= allocated);
+        }
+
+        // Grant single slot to allocation, returns true iff more slot(s) are required
+        bool grant()
+        {
+            std::unique_lock lock{mutex};
+            granted++;
+            allocated++;
+            return allocated < limit;
+            // WARNING: `waiter` iterator is invalidated after returning false
+        }
+
+        ConcurrencyControl & parent;
+        Waiters::iterator waiter; // iterator to itself in Waiters list
+
+        const Slots limit;
+
+        std::mutex mutex; // the following values must be accessed under this mutex
+        Slots allocated = 0; // allocated total (including already released)
+        Slots granted = 0; // allocated, but not yet acquired
+        Slots released = 0;
+    };
+
+public:
+    ConcurrencyControl()
+        : cur_waiter(waiters.end())
+    {}
+
+    // WARNING: all Allocation objects MUST be destructed before ConcurrencyControl
+    // NOTE: Recommended way to achieve this is to use `instance()` and do graceful shutdown of queries
+    ~ConcurrencyControl()
+    {
+        assert(waiters.empty());
+    }
+
+    // Allocate at least `min` and at most `max` slots.
+    // If not all `max` slots were successfully allocated, a subscription for later allocation is created
+    // Use Allocation::tryAcquire() to acquire allocated slot, before running a thread.
+    [[nodiscard]] AllocationPtr allocate(Slots min, Slots max)
+    {
+        return AllocationPtr(new Allocation(*this, min, max));
+    }
+
+    void setMaxConcurrency(Slots value)
+    {
+        std::unique_lock lock{mutex};
+        max_concurrency = std::max<Slots>(1, value); // never allow max_concurrency to be zero
+        schedule(lock);
+    }
+
+    static ConcurrencyControl & instance()
+    {
+        static ConcurrencyControl result;
+        return result;
+    }
+
+private:
+    Slots available(std::unique_lock<std::mutex> &)
+    {
+        if (cur_concurrency < max_concurrency)
+            return max_concurrency - cur_concurrency;
+        else
+            return 0;
+    }
+
+    // Round-robin scheduling of available slots among waiting allocations
+    void schedule(std::unique_lock<std::mutex> &)
+    {
+        while (cur_concurrency < max_concurrency && !waiters.empty())
+        {
+            cur_concurrency++;
+            if (cur_waiter == waiters.end())
+                cur_waiter = waiters.begin();
+            Allocation * allocation = *cur_waiter;
+            if (allocation->grant())
+                ++cur_waiter;
+            else
+                waiters.erase(cur_waiter++); // last required slot has just been granted -- stop waiting
+        }
+    }
+
+    std::mutex mutex;
+    Waiters waiters;
+    Waiters::iterator cur_waiter; // round-robin pointer
+    Slots max_concurrency = Slots(-1);
```

Comment:
std::numeric_limits<Slots>::max  ?
Alternatively, std::thread ::hardware_concurrency() , while it is rather problematic option, e.g. it can return 0.